src/sources/http_scrape/mod.rs

```diff
@@ -0,0 +1,212 @@
+//! Common logic for sources that are HTTP scrapers.
+//!
+//! Specific HTTP scraping sources will:
+//!   - Call get_url() to build the URL(s) to scrape.
+//!   - Implmement a specific context struct which:
+//!       - Contains the data that source needs in order to process the HTTP responses into internal_events
+//!       - Implements the HttpScraper trait
+//!   - Call http_scrape() supplying the generic inputs for scraping and the source-specific
+//!     context.
+
+#[cfg(feature = "sources-http_scrape")]
+pub mod scrape;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(all(test, feature = "http-scrape-integration-tests"))]
+mod integration_tests;
+
+pub use scrape::HttpScrapeConfig;
+
+use bytes::Bytes;
+use futures_util::{stream, FutureExt, StreamExt, TryFutureExt};
+use http::{response::Parts, Uri};
+use hyper::{Body, Request};
+use std::time::{Duration, Instant};
+use std::{collections::HashMap, future::ready};
+use tokio_stream::wrappers::IntervalStream;
+
+use crate::{
+    http::{Auth, HttpClient},
+    internal_events::{
+        EndpointBytesReceived, HttpScrapeEventsReceived, HttpScrapeHttpError,
+        HttpScrapeHttpResponseError, RequestCompleted, StreamClosedError,
+    },
+    tls::TlsSettings,
+    Error, SourceSender,
+};
+use vector_common::shutdown::ShutdownSignal;
+use vector_core::{config::proxy::ProxyConfig, event::Event, ByteSizeOf};
+
+/// Contains the inputs generic to any http scrape.
+pub(crate) struct GenericHttpScrapeInputs {
+    /// Array of URLs to scrape
+    pub urls: Vec<Uri>,
+    /// Interval to scrape on in seconds
+    pub interval_secs: u64,
+    /// Map of Header+Value to apply to HTTP request
+    pub headers: Option<HashMap<String, Vec<String>>>,
+    /// Content type of the HTTP request, determined by the source
+    pub content_type: String,
+    pub auth: Option<Auth>,
+    pub tls: TlsSettings,
+    pub proxy: ProxyConfig,
+    pub shutdown: ShutdownSignal,
+}
+
+/// The default interval to scrape the http endpoint if none is configured.
+pub(crate) const fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+/// Methods that allow context-specific behavior during the scraping procedure.
+pub(crate) trait HttpScraper {
+    /// (Optional) Called before the HTTP request is made, allows building context.
+    fn build(&mut self, _url: &Uri) {}
+
+    /// Called after the HTTP request succeeds and returns the decoded/parsed Event array.
+    fn on_response(&mut self, url: &Uri, header: &Parts, body: &Bytes) -> Option<Vec<Event>>;
+
+    /// (Optional) Called if the HTTP response is not 200 ('OK').
+    fn on_http_response_error(&self, _uri: &Uri, _header: &Parts) {}
+}
+
+/// Builds a url for the HTTP requests.
+pub(crate) fn get_url(uri: &Uri, query: &Option<HashMap<String, Vec<String>>>) -> Uri {
```

Comment:
I'm wondering if this could/should be shared with other components that have configurable query parameters?

Reply:
I could drop it in `sources/util/http/query.rs` , unless there is a better place?

Reply:
I did a quick search, so far haven't found other usages of query params like this, but I might have missed it.

Reply:
I do see other components utilizing query, but they are building the urls differently it seems.
It does however seem like something that might be worth standardizing.